File: game_models.py
"""
Game Models and Data Structures
Contains all the core data classes and enums for the BC Forestry Simulator
"""

from dataclasses import dataclass, field
from typing import List, Dict, Tuple
from enum import Enum


class PermitStatus(Enum):
    PENDING = "pending"
    APPROVED = "approved"
    DENIED = "denied"
    DELAYED = "delayed"


class CertificationType(Enum):
    FSC = "FSC"  # Forest Stewardship Council
    PEFC = "PEFC"  # Programme for Endorsement of Forest Certification
    SFI = "SFI"  # Sustainable Forestry Initiative


class DisasterType(Enum):
    BEETLE_KILL = "beetle_kill"
    WILDFIRE = "wildfire" 
    WINDSTORM = "windstorm"
    DROUGHT = "drought"
    FLOOD = "flood"


@dataclass
class HarvestBlock:
    id: str
    volume_m3: int
    year_planned: int
    permit_status: PermitStatus = PermitStatus.PENDING
    permit_submitted: int = 0
    processing_time: int = 0
    first_nations_agreement: bool = False
    heritage_cleared: bool = False
    old_growth_affected: bool = False
    priority: int = 1  # 1=low, 2=medium, 3=high priority for permitting
    disaster_affected: bool = False
    disaster_type: DisasterType = None
    volume_loss_percent: float = 0.0


@dataclass
class FirstNation:
    name: str
    relationship_level: float = 0.5  # 0-1 scale
    treaty_area: bool = False
    active_negotiations: bool = False
    agreement_signed: bool = False
    consultation_cost: int = 8000
    last_consultation_year: int = 0
    consultation_frequency_required: int = 2  # Years between required consultations
    
    def needs_consultation(self, current_year: int) -> bool:
        return (current_year - self.last_consultation_year) >= self.consultation_frequency_required


@dataclass
class Certification:
    cert_type: CertificationType
    obtained_year: int = 0
    annual_cost: int = 25000
    revenue_bonus: float = 0.15  # 15% premium on certified timber
    reputation_bonus: float = 0.1
    active: bool = False


@dataclass
class GameState:
    year: int = 2025
    quarter: int = 1  # Q1, Q2, Q3, Q4
    budget: int = 2500000
    reputation: float = 0.5
    survival_bonus: float = 0.0
    permit_bonus: float = 0.0
    region: str = ""
    species: str = ""
    prep: int = 0
    training: int = 0
    
    # Enhanced game elements
    annual_allowable_cut: int = 150000  # m3/year
    aac_decline_rate: float = 0.03  # 3% per year
    harvest_blocks: List[HarvestBlock] = field(default_factory=list)
    first_nations: List[FirstNation] = field(default_factory=list)
    certifications: List[Certification] = field(default_factory=list)
    mill_capacity: int = 100000  # m3/year
    jobs_created: int = 0
    biodiversity_score: float = 0.5
    cumulative_disturbance: float = 0.0  # hectares
    disturbance_cap: float = 50000  # Blueberry River style cap
    
    # Policy tracking
    glyphosate_banned: bool = False
    old_growth_deferrals_expanded: bool = False
    permit_backlog_days: int = 120  # Current government target: 25 days for wildfire, 40+ for others
    
    # Economic tracking
    revenue_per_m3: int = 85  # CAD
    operating_cost_per_m3: int = 45
    total_revenue: int = 0
    total_costs: int = 0
    
    # Win condition tracking
    years_operated: int = 0
    consecutive_profitable_years: int = 0
    social_license_maintained: bool = True
    
    # First Nations liaison and CEO
    fn_liaison: object = None  # Will hold LiaisonType object
    ceo: object = None  # Will hold CEOProfile object
    quarterly_profit: int = 0  # Tracks quarterly profit for CEO profit sharing
    
    def get_active_certifications(self) -> List[Certification]:
        return [c for c in self.certifications if c.active]
    
    def get_certification_revenue_bonus(self) -> float:
        return sum(c.revenue_bonus for c in self.get_active_certifications())
    
    def get_certification_reputation_bonus(self) -> float:
        return sum(c.reputation_bonus for c in self.get_active_certifications())

File: js/gameModels.js
/**
 * Game Models and Data Structures
 * Contains all the core data classes and enums for the BC Forestry Simulator
 */

export const PermitStatus = {
  PENDING: "pending",
  APPROVED: "approved",
  DENIED: "denied",
  DELAYED: "delayed",
};

export const CertificationType = {
  FSC: "FSC", // Forest Stewardship Council
  PEFC: "PEFC", // Programme for Endorsement of Forest Certification
  SFI: "SFI", // Sustainable Forestry Initiative
};

export const DisasterType = {
  BEETLE_KILL: "beetle_kill",
  WILDFIRE: "wildfire",
  WINDSTORM: "windstorm",
  DROUGHT: "drought",
  FLOOD: "flood",
};

export class HarvestBlock {
  constructor({
    id,
    volume_m3,
    year_planned,
    permit_status = PermitStatus.PENDING,
    permit_submitted = 0,
    processing_time = 0,
    first_nations_agreement = false,
    heritage_cleared = false,
    old_growth_affected = false,
    priority = 1, // 1=low, 2=medium, 3=high
    disaster_affected = false,
    disaster_type = null,
    volume_loss_percent = 0.0,
    fn_consulted = false,
    log_grade_distribution = { sawlogs: 0.6, pulp: 0.3, firewood: 0.1 },
  }) {
    this.id = id;
    this.volume_m3 = volume_m3;
    this.year_planned = year_planned;
    this.permit_status = permit_status;
    this.permit_submitted = permit_submitted;
    this.processing_time = processing_time;
    this.first_nations_agreement = first_nations_agreement;
    this.heritage_cleared = heritage_cleared;
    this.old_growth_affected = old_growth_affected;
    this.priority = priority;
    this.disaster_affected = disaster_affected;
    this.disaster_type = disaster_type;
    this.volume_loss_percent = volume_loss_percent;
    this.fn_consulted = fn_consulted;
    this.log_grade_distribution = log_grade_distribution;
  }
}

export class FirstNation {
  constructor({
    name,
    relationship_level = 0.5, // 0-1 scale
    treaty_area = false,
    active_negotiations = false,
    agreement_signed = false,
    consultation_cost = 8000,
    last_consultation_year = 0,
    consultation_frequency_required = 2, // Years
  }) {
    this.name = name;
    this.relationship_level = relationship_level;
    this.treaty_area = treaty_area;
    this.active_negotiations = active_negotiations;
    this.agreement_signed = agreement_signed;
    this.consultation_cost = consultation_cost;
    this.last_consultation_year = last_consultation_year;
    this.consultation_frequency_required = consultation_frequency_required;
  }

  needs_consultation(current_year) {
    return (
      current_year - this.last_consultation_year >=
      this.consultation_frequency_required
    );
  }
}

export class Certification {
  constructor({
    cert_type,
    obtained_year = 0,
    annual_cost = 25000,
    revenue_bonus = 0.15,
    reputation_bonus = 0.1,
    active = false,
  }) {
    this.cert_type = cert_type;
    this.obtained_year = obtained_year;
    this.annual_cost = annual_cost;
    this.revenue_bonus = revenue_bonus;
    this.reputation_bonus = reputation_bonus;
    this.active = active;
  }
}

export class GameState {
  constructor() {
    this.year = 2025;
    this.quarter = 1; // Q1, Q2, Q3, Q4
    this.budget = 2500000;
    this.reputation = 0.5;
    this.survival_bonus = 0.0;
    this.permit_bonus = 0.0;
    this.region = "";
    this.species = "";
    this.prep = 0;
    this.training = 0;

    // Enhanced game elements
    this.annual_allowable_cut = 150000; // m3/year
    this.aac_decline_rate = 0.03; // 3% per year
    this.harvest_blocks = [];
    this.first_nations = [];
    this.certifications = [];
    this.mill_capacity = 100000; // m3/year
    this.jobs_created = 0;
    this.biodiversity_score = 0.5;
    this.cumulative_disturbance = 0.0; // hectares
    this.disturbance_cap = 50000; // Blueberry River style cap

    // Policy tracking
    this.glyphosate_banned = false;
    this.old_growth_deferrals_expanded = false;
    this.permit_backlog_days = 120;

    // Economic tracking
    this.log_prices = {
      sawlogs: 120, // $/m3
      pulp: 60,     // $/m3
      firewood: 30  // $/m3
    };
    this.operating_cost_per_m3 = 45;
    this.total_revenue = 0;
    this.total_costs = 0;

    // Win condition tracking
    this.years_operated = 0;
    this.consecutive_profitable_years = 0;
    this.social_license_maintained = true;

    // Community & story progression
    this.community_support = 0.5; // 0-1 scale
    this.story_stage = 0;
    this.story_branch = null;

    // First Nations liaison and CEO
    this.fn_liaison = null;
    this.ceo = null;
    this.quarterly_profit = 0;
    
    // Safety and criminal tracking
    this.safety_violations = 0;
    this.safety_fatalities = 0;
    this.equipment_condition = 1.0; // 0-1 scale
    this.under_criminal_investigation = false;
    this.criminal_convictions = 0;
    
    // Company name
    this.companyName = "Northern Forest Solutions";
  }

  get_active_certifications() {
    return this.certifications.filter((c) => c.active);
  }

  get_certification_revenue_bonus() {
    return this.get_active_certifications().reduce(
      (total, c) => total + c.revenue_bonus,
      0
    );
  }

  get_certification_reputation_bonus() {
    return this.get_active_certifications().reduce(
      (total, c) => total + c.reputation_bonus,
      0
    );
  }
}
